classes/local/store/object_file_system.php

```diff
@@ -804,13 +806,61 @@ protected function copy_from_external_to_local($contenthash) {
      */
     public function redirect_to_presigned_url($contenthash, $headers = array()) {
         try {
-            redirect($this->externalclient->generate_presigned_url($contenthash, $headers));
+            redirect($this->embed_origin_url(
+                $this->externalclient->generate_presigned_url($contenthash, $headers),
+            ));
         } catch (\Exception $e) {
             debugging('Failed to redirect to pre-signed url: ' . $e->getMessage());
             return false;
         }
     }
 
+    /**
+     * Embed origin URL in presigned URL.
+     *
+     * @param string $presignedurl
+     * @return string $presignedurl with embedded origin
+     * @throws moodle_exception
+     */
+    public function embed_origin_url($presignedurl): string {
+        $me = me();
+        if (empty($me)) {
+            throw new moodle_exception('invalidurl');
+        }
+
+        return sprintf(
+            "%s#%s%s",
+            $presignedurl,
+            self::PLUGINFILE_ORIGIN_SEGMENT_PREFIX,
+            urlencode($me)
+        );
+    }
+
+    /**
+     * Convert redirected URLs in $text.
+     *
+     * @param string  $text The content that may contain ULRs in need of rewriting.
+     * @return string The processed text.
+     */
+    public function rewrite_pluginfile_urls($text): string {
+        if (!$this->presigned_url_configured()) { // Do nothing.
+            return $text;
+        }
+
+        $re = sprintf(
+            '!%s\S*?\#%s(\S+\w)!',
+            preg_quote($this->externalclient->presigned_domain()),
```

Comment:
I'm just wondering if we can be more relaxed here and accept anything that looks like a url and has a fragment in the right shape. It would then work in a variety of edge cases like cutting and pasting urls between environments which either do or do no have signing on, or do have signing on with different domains. In all cases their local moodle urls should all have the same local part and Just Work (tm)

Reply:
I actually started like that, but then thought we may want to limit the conversion. If we want to relax it then we need to embed full urls, not just paths to be glued to wwwroot. I suppose we can do that, and let acl decides what's available.